File: foundational_design_patterns/7_multi_agent_collaboration/src/peer_swarm.py
"""
Multi-Agent Collaboration: Peer/Swarm topology

Anchor scenario: support-ops triage with peer negotiation.

Preferred runtime in 2026:
- langgraph-swarm (if installed)
Fallback:
- explicit peer loop with critique/synthesis
"""

from __future__ import annotations

import os
import sys
from dataclasses import dataclass, field

from pathlib import Path

ROOT_DIR = next(parent for parent in Path(__file__).resolve().parents if (parent / "ssl_fix.py").exists())
if str(ROOT_DIR) not in sys.path:
    sys.path.insert(0, str(ROOT_DIR))

from repo_support import configure_example, get_default_model

configure_example(__file__)

from langchain_openai import ChatOpenAI


@dataclass
class PeerState:
    problem: str
    analyst_notes: list[str] = field(default_factory=list)
    final_decision: str = ""


llm = ChatOpenAI(model=get_default_model(), temperature=0)


def peer_analysis(role: str, problem: str) -> str:
    return llm.invoke(
        f"You are {role}. Analyze this problem and propose one actionable recommendation:\n{problem}"
    ).content


def peer_critique(notes: list[str]) -> str:
    prompt = f"""
You are a peer reviewer. Critique these peer recommendations and resolve conflicts.

{notes}

Return: conflict_summary + merged recommendation.
"""
    return llm.invoke(prompt).content


def try_swarm_package() -> bool:
    try:
        import langgraph_swarm  # noqa: F401

        print("✅ langgraph-swarm detected. (This demo uses portable fallback loop.)")
        return True
    except Exception:
        print("ℹ️ langgraph-swarm not installed; using fallback peer loop.")
        return False


def run_peer_swarm(problem: str) -> PeerState:
    state = PeerState(problem=problem)
    roles = ["SRE analyst", "Support lead", "Product reliability manager"]

    for role in roles:
        state.analyst_notes.append(f"[{role}]\n{peer_analysis(role, problem)}")

    state.final_decision = peer_critique(state.analyst_notes)
    return state


def main() -> None:
    print("\n" + "=" * 80)
    print("MULTI-AGENT — PEER/SWARM")
    print("=" * 80)

    if not os.getenv("OPENAI_API_KEY"):
        print("❌ OPENAI_API_KEY is required.")
        return

    try_swarm_package()

    problem = "Multiple enterprise accounts report payment approval delays after release v2.8.4."
    state = run_peer_swarm(problem)

    print("\n🤝 Peer notes:")
    for note in state.analyst_notes:
        print(note)
        print("-" * 60)

    print("\n✅ Final peer decision:")
    print(state.final_decision)


if __name__ == "__main__":
    main()

File: foundational_design_patterns/7_multi_agent_collaboration/typescript/src/peer_swarm.ts
/**
 * Multi-Agent Collaboration: Peer/Swarm topology
 * — TypeScript port of src/peer_swarm.py.
 */

import { ChatOpenAI } from "@langchain/openai";

const MODEL = process.env.OPENAI_MODEL ?? "gpt-4o-mini";

export interface PeerState {
  problem: string;
  analyst_notes: string[];
  final_decision: string;
}

const llm = new ChatOpenAI({
  model: MODEL,
  temperature: 0,
});

export async function peerAnalysis(
  role: string,
  problem: string,
): Promise<string> {
  const response = await llm.invoke(
    `You are ${role}. Analyze this problem and propose one actionable recommendation:\n${problem}`,
  );
  return String(response.content ?? "");
}

export async function peerCritique(notes: string[]): Promise<string> {
  const prompt = `
You are a peer reviewer. Critique these peer recommendations and resolve conflicts.

${notes.join("\n")}

Return: conflict_summary + merged recommendation.
`;

  const response = await llm.invoke(prompt);
  return String(response.content ?? "");
}

export async function trySwarmPackage(): Promise<boolean> {
  try {
    const importer = new Function(
      "specifier",
      "return import(specifier);",
    ) as (specifier: string) => Promise<unknown>;
    await importer("langgraph-swarm");
    console.log(
      "✅ langgraph-swarm detected. (This demo uses portable fallback loop.)",
    );
    return true;
  } catch {
    console.log("ℹ️ langgraph-swarm not installed; using fallback peer loop.");
    return false;
  }
}

export async function runPeerSwarm(problem: string): Promise<PeerState> {
  const state: PeerState = {
    problem,
    analyst_notes: [],
    final_decision: "",
  };
  const roles = [
    "SRE analyst",
    "Support lead",
    "Product reliability manager",
  ];

  for (const role of roles) {
    state.analyst_notes.push(
      `[${role}]\n${await peerAnalysis(role, problem)}`,
    );
  }

  state.final_decision = await peerCritique(state.analyst_notes);
  return state;
}

async function main(): Promise<void> {
  console.log("\n" + "=".repeat(80));
  console.log("MULTI-AGENT — PEER/SWARM");
  console.log("=".repeat(80));

  if (!process.env.OPENAI_API_KEY) {
    console.log("❌ OPENAI_API_KEY is required.");
    return;
  }

  await trySwarmPackage();

  const problem =
    "Multiple enterprise accounts report payment approval delays after release v2.8.4.";
  const state = await runPeerSwarm(problem);

  console.log("\n🤝 Peer notes:");
  for (const note of state.analyst_notes) {
    console.log(note);
    console.log("-".repeat(60));
  }

  console.log("\n✅ Final peer decision:");
  console.log(state.final_decision);
}

if (import.meta.main) {
  await main();
}
